File: lib/core/services/news_scraping_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:html/parser.dart' as html_parser;
import 'package:html/dom.dart' as html_dom;

class NewsArticle {
  final String title;
  final String content;
  final String url;
  final String source;
  final DateTime publishedAt;
  final String? imageUrl;

  NewsArticle({
    required this.title,
    required this.content,
    required this.url,
    required this.source,
    required this.publishedAt,
    this.imageUrl,
  });
}

class NewsScrapingService {
  static const List<String> nflSources = [
    'https://www.nfl.com/news',
    'https://www.espn.com/nfl/',
    'https://www.cbssports.com/nfl/',
    'https://www.si.com/nfl',
  ];

  static const Map<String, Map<String, String>> sourceSelectors = {
    'nfl.com': {
      'articleSelector': '.nfl-c-article__content',
      'titleSelector': 'h1, .nfl-c-article__title',
      'contentSelector': '.nfl-c-article__body p',
      'linkSelector': 'a[href*="/news/"]',
    },
    'espn.com': {
      'articleSelector': '.contentItem__content',
      'titleSelector': 'h1, .contentItem__title',
      'contentSelector': '.article-body p',
      'linkSelector': 'a[href*="/nfl/"]',
    },
    'cbssports.com': {
      'articleSelector': '.ArticleBody',
      'titleSelector': 'h1, .ArticleTitle',
      'contentSelector': '.ArticleBody p',
      'linkSelector': 'a[href*="/nfl/"]',
    },
    'si.com': {
      'articleSelector': '.article-content',
      'titleSelector': 'h1, .article-title',
      'contentSelector': '.article-content p',
      'linkSelector': 'a[href*="/nfl/"]',
    },
  };

  Future<List<NewsArticle>> scrapeLatestNews() async {
    final List<NewsArticle> allArticles = [];
    
    for (final source in nflSources) {
      try {
        print('Scraping news from: $source');
        final articles = await _scrapeSource(source);
        allArticles.addAll(articles);
        print('Found ${articles.length} articles from $source');
      } catch (e) {
        print('Error scraping $source: $e');
      }
    }

    // Filter articles to only include those from the last 6 hours
    final sixHoursAgo = DateTime.now().subtract(const Duration(hours: 6));
    final recentArticles = allArticles
        .where((article) => article.publishedAt.isAfter(sixHoursAgo))
        .toList();

    // Sort by publication date (newest first)
    recentArticles.sort((a, b) => b.publishedAt.compareTo(a.publishedAt));

    print('Total recent articles found: ${recentArticles.length}');
    return recentArticles;
  }

  Future<List<NewsArticle>> _scrapeSource(String baseUrl) async {
    final client = http.Client();
    try {
      final response = await client.get(
        Uri.parse(baseUrl),
        headers: {
          'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
        },
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to fetch $baseUrl: ${response.statusCode}');
      }

      final document = html_parser.parse(response.body);
      final domain = _extractDomain(baseUrl);
      final selectors = sourceSelectors[domain] ?? sourceSelectors['nfl.com']!;

      final articles = <NewsArticle>[];
      
      // Find article links
      final articleLinks = document.querySelectorAll(selectors['linkSelector']!);
      
      for (final link in articleLinks.take(10)) { // Limit to 10 articles per source
        final href = link.attributes['href'];
        if (href == null) continue;

        final articleUrl = _resolveUrl(baseUrl, href);
        try {
          final article = await _scrapeArticle(articleUrl, selectors);
          if (article != null) {
            articles.add(article);
          }
        } catch (e) {
          print('Error scraping article $articleUrl: $e');
        }
      }

      return articles;
    } finally {
      client.close();
    }
  }

  Future<NewsArticle?> _scrapeArticle(String url, Map<String, String> selectors) async {
    final client = http.Client();
    try {
      final response = await client.get(
        Uri.parse(url),
        headers: {
          'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
        },
      );

      if (response.statusCode != 200) {
        return null;
      }

      final document = html_parser.parse(response.body);
      
      // Extract title
      final titleElement = document.querySelector(selectors['titleSelector']!);
      if (titleElement == null) return null;
      final title = titleElement.text.trim();
      if (title.isEmpty) return null;

      // Extract content
      final contentElements = document.querySelectorAll(selectors['contentSelector']!);
      final content = contentElements
          .map((e) => e.text.trim())
          .where((text) => text.isNotEmpty)
          .join('\n\n');

      if (content.isEmpty) return null;

      // Extract image URL
      final imageElement = document.querySelector('img');
      final imageUrl = imageElement?.attributes['src'];

      // Extract publication date (try multiple selectors)
      DateTime publishedAt = DateTime.now();
      final dateSelectors = [
        'time[datetime]',
        '.published-date',
        '.article-date',
        '[data-published]',
      ];

      for (final selector in dateSelectors) {
        final dateElement = document.querySelector(selector);
        if (dateElement != null) {
          final dateString = dateElement.attributes['datetime'] ?? 
                           dateElement.attributes['data-published'] ??
                           dateElement.text.trim();
          try {
            publishedAt = DateTime.parse(dateString);
            break;
          } catch (e) {
            // Try parsing relative dates
            if (dateString.contains('ago')) {
              publishedAt = _parseRelativeDate(dateString);
              break;
            }
          }
        }
      }

      return NewsArticle(
        title: title,
        content: content,
        url: url,
        source: _extractDomain(url),
        publishedAt: publishedAt,
        imageUrl: imageUrl,
      );
    } finally {
      client.close();
    }
  }

  String _extractDomain(String url) {
    final uri = Uri.parse(url);
    final host = uri.host.toLowerCase();
    if (host.contains('nfl.com')) return 'nfl.com';
    if (host.contains('espn.com')) return 'espn.com';
    if (host.contains('cbssports.com')) return 'cbssports.com';
    if (host.contains('si.com')) return 'si.com';
    return host;
  }

  String _resolveUrl(String baseUrl, String href) {
    if (href.startsWith('http')) return href;
    if (href.startsWith('//')) return 'https:$href';
    if (href.startsWith('/')) {
      final baseUri = Uri.parse(baseUrl);
      return '${baseUri.scheme}://${baseUri.host}$href';
    }
    return '$baseUrl/$href';
  }

  DateTime _parseRelativeDate(String dateString) {
    final now = DateTime.now();
    if (dateString.contains('minute')) {
      final minutes = int.tryParse(dateString.split(' ')[0]) ?? 0;
      return now.subtract(Duration(minutes: minutes));
    } else if (dateString.contains('hour')) {
      final hours = int.tryParse(dateString.split(' ')[0]) ?? 0;
      return now.subtract(Duration(hours: hours));
    } else if (dateString.contains('day')) {
      final days = int.tryParse(dateString.split(' ')[0]) ?? 0;
      return now.subtract(Duration(days: days));
    }
    return now;
  }
}

File: railway-backend/src/services/newsScrapingService.js
const axios = require('axios');
const cheerio = require('cheerio');

class NewsScrapingService {
  constructor() {
    this.sources = [
      {
        name: 'ESPN NFL',
        url: 'https://www.espn.com/nfl/news',
        selector: '.headlineStack__list .contentItem',
        titleSelector: '.contentItem__title a',
        linkSelector: '.contentItem__title a',
        contentSelector: '.contentItem__subhead',
        baseUrl: 'https://www.espn.com'
      },
      {
        name: 'NFL.com',
        url: 'https://www.nfl.com/news',
        selector: '.nfl-c-card-grid__item',
        titleSelector: '.nfl-c-card__title a',
        linkSelector: '.nfl-c-card__title a',
        contentSelector: '.nfl-c-card__summary',
        baseUrl: 'https://www.nfl.com'
      },
      {
        name: 'CBS Sports NFL',
        url: 'https://www.cbssports.com/nfl/news',
        selector: '.Article',
        titleSelector: '.Article-title a',
        linkSelector: '.Article-title a',
        contentSelector: '.Article-summary',
        baseUrl: 'https://www.cbssports.com'
      }
    ];
    
    this.axiosConfig = {
      timeout: 10000,
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
      }
    };
  }

  async scrapeLatestNews() {
    const allArticles = [];
    
    for (const source of this.sources) {
      try {
        console.log(`🔍 Scraping ${source.name}...`);
        const articles = await this.scrapeSource(source);
        allArticles.push(...articles);
        console.log(`✅ Found ${articles.length} articles from ${source.name}`);
        
        // Add delay between sources to be respectful
        await this.delay(2000);
      } catch (error) {
        console.error(`❌ Error scraping ${source.name}:`, error.message);
      }
    }
    
    // Remove duplicates and filter recent articles
    const uniqueArticles = this.removeDuplicates(allArticles);
    const recentArticles = this.filterRecentArticles(uniqueArticles);
    
    console.log(`📊 Total unique recent articles: ${recentArticles.length}`);
    return recentArticles;
  }

  async scrapeSource(source) {
    try {
      const response = await axios.get(source.url, this.axiosConfig);
      const $ = cheerio.load(response.data);
      const articles = [];

      $(source.selector).each((index, element) => {
        try {
          // Limit to first 10 articles per source
          if (index >= 10) return false;

          const $element = $(element);
          const title = $element.find(source.titleSelector).text().trim();
          const relativeLink = $element.find(source.linkSelector).attr('href');
          const content = $element.find(source.contentSelector).text().trim();

          if (!title || !relativeLink) return;

          // Build full URL
          let fullUrl = relativeLink;
          if (relativeLink.startsWith('/')) {
            fullUrl = source.baseUrl + relativeLink;
          } else if (!relativeLink.startsWith('http')) {
            fullUrl = source.baseUrl + '/' + relativeLink;
          }

          // Generate a simple ID from URL
          const id = this.generateArticleId(fullUrl);

          articles.push({
            id,
            title,
            content: content || title, // Use title as fallback content
            url: fullUrl,
            source: source.name,
            publishedAt: new Date(),
            imageUrl: this.generatePlaceholderImage(index)
          });
        } catch (error) {
          console.error(`Error processing article element:`, error.message);
        }
      });

      return articles;
    } catch (error) {
      console.error(`Error fetching ${source.url}:`, error.message);
      return [];
    }
  }

  generateArticleId(url) {
    // Simple hash function for URL
    let hash = 0;
    for (let i = 0; i < url.length; i++) {
      const char = url.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32-bit integer
    }
    return Math.abs(hash).toString();
  }

  generatePlaceholderImage(index) {
    // Generate placeholder images for articles
    const imageIds = [100, 101, 102, 103, 104, 105, 106, 107, 108, 109];
    const imageId = imageIds[index % imageIds.length];
    return `https://picsum.photos/400/200?random=${imageId}`;
  }

  removeDuplicates(articles) {
    const seen = new Set();
    return articles.filter(article => {
      const key = article.title.toLowerCase().replace(/[^a-z0-9]/g, '');
      if (seen.has(key)) {
        return false;
      }
      seen.add(key);
      return true;
    });
  }

  filterRecentArticles(articles) {
    const sixHoursAgo = new Date(Date.now() - 6 * 60 * 60 * 1000);
    return articles.filter(article => {
      // For scraped articles, we assume they're recent
      // In a more sophisticated implementation, we'd parse publish dates
      return true;
    });
  }

  delay(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

module.exports = NewsScrapingService;
